ssdeep.go

```diff
@@ -137,66 +112,142 @@ type fuzzyReader interface {
+		return "", errors.New("Did not process files large enough to produce meaningful results")
 	}
 
-	sdeep.GetBlockSize(n)
+	state.getBlockSize(n)
 	for {
 		f.Seek(0, 0)
 		r := bufio.NewReader(f)
-		sdeep.process(r)
-		if sdeep.blockSize < blockMin {
-			return nil, errors.New("Unable to establish a sufficient block size")
+		state.process(r)
+		if state.blockSize < blockMin {
+			return "", errors.New("Unable to establish a sufficient block size")
 		}
-		if len(sdeep.hashString1) < spamSumLength/2 {
-			sdeep.blockSize = sdeep.blockSize / 2
-			sdeep.blockHash1 = hashInit
-			sdeep.blockHash2 = hashInit
-			sdeep.hashString1 = ""
-			sdeep.hashString2 = ""
+		if len(state.hashString1) < spamSumLength/2 {
+			state.blockSize = state.blockSize / 2
+			state.blockHash1 = hashInit
+			state.blockHash2 = hashInit
+			state.hashString1 = ""
+			state.hashString2 = ""
 		} else {
-			rh := sdeep.rollingState.rollSum()
+			rh := state.rollingState.rollSum()
 			if rh != 0 {
 				// Finalize the hash string with the remaining data
-				sdeep.hashString1 += string(b64[sdeep.blockHash1%64])
-				sdeep.hashString2 += string(b64[sdeep.blockHash2%64])
+				state.hashString1 += string(b64[state.blockHash1%64])
+				state.hashString2 += string(b64[state.blockHash2%64])
 			}
 			break
 		}
 	}
-	return &FuzzyHash{
-		blockSize:    sdeep.blockSize,
-		hashString1:  sdeep.hashString1,
-		hashString2:  sdeep.hashString2,
-		fileLocation: fileLocation,
-	}, nil
+	return fmt.Sprintf("%d:%s:%s", state.blockSize, state.hashString1, state.hashString2), nil
 }
 
-// FuzzyFile hash of a provided reader
-func (sdeep *SSDEEP) FuzzyFile(f *os.File, fileLocation string) (*FuzzyHash, error) {
-	n, err := GetFileSize(f)
+// FuzzyFilename computes the fuzzy hash of a file.
+// FuzzyFilename will opens, reads, and hashes the contents of the file 'filename'.
+// It is the caller's responsibility to append the filename to the result after computation.
+// Returns an error when the file doesn't exist or ssdeep could not be computed on the file.
+func FuzzyFilename(filename string) (string, error) {
+	f, err := os.Open(filename)
 	if err != nil {
-		return nil, err
+		return "", err
 	}
+	defer f.Close()
 
-	return sdeep.fuzzyReader(f, n, fileLocation)
+	return FuzzyFile(f)
 }
 
-// FuzzyByte hash of a provided byte array
-func (sdeep *SSDEEP) FuzzyByte(blob []byte) (*FuzzyHash, error) {
-	n := len(blob)
-	br := bytes.NewReader(blob)
+// FuzzyFile computes the fuzzy hash of a file using os.File pointer.
+// FuzzyFile will computes the fuzzy hash of the contents of the open file, starting at the beginning of the file.
+// When finished, the file pointer is returned to its original position.
+// If an error occurs, the file pointer's value is undefined.
+// It is the callers's responsibility to append the filename to the result after computation.
+// Returns an error when ssdeep could not be computed on the file.
+func FuzzyFile(f *os.File) (string, error) {
```

Comment:
@davidt99 did you think about this? IMO  we should try to finalize the API before we do a 1.0 release.

Reply:
Didn't notice that. It's a good idea to expose `FuzzyReader`, but one concern I have is it might not be intuitive to the caller to pass `*os.File` to a function that have `FuzzyReader` as a parameter. What do you say about adding new API that accepts `FuzzyReader`?

Reply:
Like this one? https://github.com/intezer/ssdeep/blob/19cb84b1d6ec7f1b00bcf4a0d091a2d7328b4489/ssdeep.go#L124
We should also make a distinction between the type fuzzyReader and the method fuzzyReader

Reply:
This one exactly, I'll rename the struct then